File: robot/ros_ws/src/modules/asm_mighty/mighty/src/mighty/convert_vicon_to_state.cpp
#include "mighty/convert_vicon_to_state.hpp"

PoseTwistToStateNode::PoseTwistToStateNode() : Node("pose_twist_to_state_node") {
  // define quality of service: all messages that you want to receive must have the same
  rmw_qos_profile_t custom_qos_profile = rmw_qos_profile_default;
  custom_qos_profile.depth = 1;
  custom_qos_profile.reliability =
      rmw_qos_reliability_policy_t::RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT;
  custom_qos_profile.history = rmw_qos_history_policy_t::RMW_QOS_POLICY_HISTORY_KEEP_LAST;
  custom_qos_profile.durability =
      rmw_qos_durability_policy_t::RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL;

  sub_pose_.subscribe(this, "world", custom_qos_profile);
  sub_twist_.subscribe(this, "twist", custom_qos_profile);

  // Time synchronizer setup
  pose_twist_sync_.reset(new Sync(MySyncPolicy(10), sub_pose_, sub_twist_));
  pose_twist_sync_->registerCallback(std::bind(&PoseTwistToStateNode::callback, this,
                                               std::placeholders::_1, std::placeholders::_2));

  // Publisher setup — depth 1: only the latest state matters.
  rclcpp::QoS state_qos(rclcpp::KeepLast(1));
  state_qos.reliable().durability_volatile();
  pub_state_ = this->create_publisher<dynus_interfaces::msg::State>("state", state_qos);
}

void PoseTwistToStateNode::callback(
    const std::shared_ptr<const geometry_msgs::msg::PoseStamped>& pose_msg,
    const std::shared_ptr<const geometry_msgs::msg::TwistStamped>& twist_msg) {
  // Construct the State message
  auto state_msg = dynus_interfaces::msg::State();
  state_msg.header.stamp = this->get_clock()->now();
  state_msg.header.frame_id = pose_msg->header.frame_id;

  // Set position from PoseStamped
  state_msg.pos.x = pose_msg->pose.position.x;
  state_msg.pos.y = pose_msg->pose.position.y;
  state_msg.pos.z = pose_msg->pose.position.z;

  // Set velocity from TwistStamped
  state_msg.vel.x = twist_msg->twist.linear.x;
  state_msg.vel.y = twist_msg->twist.linear.y;
  state_msg.vel.z = twist_msg->twist.linear.z;

  // Set orientation from PoseStamped
  state_msg.quat.x = pose_msg->pose.orientation.x;
  state_msg.quat.y = pose_msg->pose.orientation.y;
  state_msg.quat.z = pose_msg->pose.orientation.z;
  state_msg.quat.w = pose_msg->pose.orientation.w;

  // Publish the State message
  pub_state_->publish(state_msg);
  RCLCPP_INFO(this->get_logger(), "Published State");
}

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<PoseTwistToStateNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: robot/ros_ws/src/modules/asm_mighty/mighty/scripts/convert_vicon_to_state.py
#!/usr/bin/env python3

# /* ----------------------------------------------------------------------------
#  * Copyright 2025, Kota Kondo, Aerospace Controls Laboratory
#  * Massachusetts Institute of Technology
#  * All Rights Reserved
#  * Authors: Kota Kondo, et al.
#  * See LICENSE file for the license information
#  * -------------------------------------------------------------------------- */

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import PoseStamped, TwistStamped, Vector3, Quaternion
from std_msgs.msg import Header
from dynus_interfaces.msg import State
from message_filters import Subscriber, ApproximateTimeSynchronizer

class PoseTwistToStateNode(Node):
    def __init__(self):
        super().__init__('pose_twist_to_state_node')

        # Create subscribers using message_filters
        self.pose_sub = Subscriber(self, PoseStamped, 'world')
        self.twist_sub = Subscriber(self, TwistStamped, 'twist')

        # Synchronize pose and twist topics
        self.sync = ApproximateTimeSynchronizer(
            [self.pose_sub, self.twist_sub], queue_size=10, slop=0.1)
        self.sync.registerCallback(self.callback)

        # Publisher
        self.state_publisher = self.create_publisher(State, 'state', 10)

    def callback(self, pose_msg, twist_msg):
        # Construct the State message
        state_msg = State()
        state_msg.header = Header()
        state_msg.header.stamp = self.get_clock().now().to_msg()
        state_msg.header.frame_id = pose_msg.header.frame_id

        # Set position from PoseStamped
        state_msg.pos = Vector3(
            x=pose_msg.pose.position.x,
            y=pose_msg.pose.position.y,
            z=pose_msg.pose.position.z
        )

        # Set velocity from TwistStamped
        state_msg.vel = Vector3(
            x=twist_msg.twist.linear.x,
            y=twist_msg.twist.linear.y,
            z=twist_msg.twist.linear.z
        )

        # Set orientation from PoseStamped
        state_msg.quat = Quaternion(
            x=pose_msg.pose.orientation.x,
            y=pose_msg.pose.orientation.y,
            z=pose_msg.pose.orientation.z,
            w=pose_msg.pose.orientation.w
        )

        # Publish the State message
        self.state_publisher.publish(state_msg)
        self.get_logger().info(f'Published State: {state_msg}')

def main(args=None):
    rclpy.init(args=args)
    node = PoseTwistToStateNode()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()
